File: unionFind.py

# implement a union find set

class union_find():
    

    self.father = []

    def __init__(self, arr):
        self.arr = arr
        self.father = [ 0 for i in range(len(arr) + 1)]
        
    def find(self, val):

        if father[x] == 0:
            return x 
        father[x] = find(father[x])
        return father[x]

    def union(self, a,b):
        root_a = self.find(a)
        root_b = self.find(b)

        if root_a != root_b:
            father[root_a] = root_b

File: Java_version/unionFind.java
import java.lang.*;

public class unionFind { 

    int [] father;
    public unionFind(int n) {
        // initialize your data structure here.
        father = new int[n + 1];
    }
    private int find(int x){
        if (father[x] != 0){
            int q = find(father[x]);
            father[x] = q;
            return q;
        }
        return x;
        
    }

    public void union(int a, int b) {
        // Write your code here
        if (!query(a, b)){
            father[find(a)] = find(b);
        }
    }
        
    public boolean query(int a, int b) {
        // Write your code here
        return find(a) == find(b);
    }
}
